### ISSUE TITLE
Enas Tutorial is not up-to-date

### ISSUE THREAD

**Issue by dzk9528** [microsoft/nni#5526]
I can not run the https://github.com/microsoft/nni/blob/v2.10/examples/nas/oneshot/enas/search.py
it give me error on this one `from nni.nas.pytorch import mutables` to indicate no mutables in nni.nas.pytorch, when I change it, it give me errors because the `MutableScope`(See https://github.com/microsoft/nni/blob/c31d2574cb418acfb80c17bb2bd03531556325bd/examples/nas/oneshot/enas/micro.py#L62) is deleted since v2.9

**Environment**:
- NNI version: v2.10
- Training service (local):
- Client OS: Ubuntu20.04
- Python version: 3.10
- PyTorch version: 1.10+cu113
- Is conda/virtualenv/venv used?: No
- Is running in Docker?: Yes

**Configuration**:
   - Search space: ENAS

**Comment by matluster** [comment]
Please check the latest documents: https://nni.readthedocs.io/en/stable/tutorials/darts.html

We might need to delete the old examples at some point...

**Comment by GoWithWind2015** [comment]
> Please check the latest documents: https://nni.readthedocs.io/en/stable/tutorials/darts.html
> 
> We might need to delete the old examples at some point...

so just repace "strategy = DartsStrategy" with  "strategy = DartsStrategy"  will be ok?

**Comment by GoWithWind2015** [comment]
is there an example showing useage of enas  that can run without modification now?

**Comment by ultmaster** [comment]
> so just repace "strategy = DartsStrategy" with  "strategy = ENAStrategy"  will be ok?

I suppose so.

**Comment by ultmaster** [comment]
> is there any example showing useage of enas  that can run without modification now?

Maybe you can find one in unit-tests. But they are not meant for beginners.

**Comment by dzk9528** [comment]
> 

I think if I follow the tutorial, then the search space need to be changed as well. But in ENAS search space, it is hard for me to set up `ENAS` for there is no instruction to build micro or macro search space

**Comment by dzk9528** [comment]
Up to now, I can only follow dart tutorial to run one shot NAS. I need to have `torchmetrics == 0.10` and `numpy == 1.23.5` for `nni == 2.10`. Enas seems to need have some controller settings as well or use `EnasTrainer`

**Comment by dzk9528** [comment]
@ultmaster Thanks for the tutorial. In the example tutorial, in the end it concluded that with cutout for data transform, the final performance of trained model is better. And I wonder if the cutout should apply to NAS search and model training or just model training?

**Comment by ultmaster** [comment]
> @ultmaster  I think if I follow the tutorial, then the search space need to be changed as well. But in ENAS search space, it is hard for me to set up for there is no instructions to build micro or macro search space

So you are trying to reproduce ENAS search space? Reproducing ENAS search space and ENAS search strategy are two independent things.

**Comment by ultmaster** [comment]
> @ultmaster Thanks for the tutorial. In the example tutorial, in the end it concluded that with cutout for data transform, the final performance of trained model is better. And I wonder if the cutout should apply to NAS search and model training or just model training?

You can use cutout at training, but the original authors didn't. I suspect it's due to efficiency reasons.

**Comment by Lijiaoa** [comment]
any updates or close? @dzk9528

**Comment by dzk9528** [comment]
@Lijiaoa Well, since DARTS example work well(96.88% in the end), I think it is good enough for me at this point. I will not stick to ENAS tutorial, so it depends on @ultmaster and to update the doc, which definitely take a lot of effort and time. I think I will close this issue tomorrow.

**Comment by GoWithWind2015** [comment]
import nni
import numpy as np
from nni.retiarii.evaluator.pytorch import Classification
from torch.utils.data import SubsetRandomSampler
from torchvision import transforms
from torchvision.datasets import CIFAR10
from nni.retiarii.evaluator.pytorch import DataLoader


from nni.retiarii.experiment.pytorch import RetiariiExperiment, RetiariiExeConfig
from nni.retiarii.strategy.rl  import PolicyBasedRL
from nni.retiarii import fixed_arch

from nni.retiarii.strategy import ENAS as ENAStrategy
from nni.retiarii.hub.pytorch import ENAS as ENASpace

from nni.retiarii.strategy import DARTS as DartsStrategy
from nni.retiarii.hub.pytorch import DARTS as DartsSpace


CIFAR_MEAN = [0.49139968, 0.48215827, 0.44653124]
CIFAR_STD = [0.24703233, 0.24348505, 0.26158768]

# fast_dev_run = True
fast_dev_run = False

# strategy = DartsStrategy()
strategy = ENAStrategy(reward_metric_name='val_acc')

transform = transforms.Compose([
    transforms.RandomCrop(32, padding=4),
    transforms.RandomHorizontalFlip(),
    transforms.ToTensor(),
    transforms.Normalize(CIFAR_MEAN, CIFAR_STD),
])

train_data = nni.trace(CIFAR10)(root='./data', train=True, download=True, transform=transform)

num_samples = len(train_data)
indices = np.random.permutation(num_samples)
split = num_samples // 2

search_train_loader = DataLoader(
    train_data, batch_size=64, num_workers=6,
    sampler=SubsetRandomSampler(indices[:split]),
)

search_valid_loader = DataLoader(
    train_data, batch_size=64, num_workers=6,
    sampler=SubsetRandomSampler(indices[split:]),
)

evaluator = Classification(
    learning_rate=1e-3,
    weight_decay=1e-4,
    train_dataloaders=search_train_loader,
    val_dataloaders=search_valid_loader,
    # max_epochs=10,
    max_epochs=1,
    gpus=1,
    fast_dev_run=fast_dev_run,
)

# model_space = DartsSpace(
#     width=16,           # the initial filters (channel number) for the model
#     num_cells=8,        # the number of stacked cells in total
#     dataset='cifar'     # to give a hint about input resolution, here is 32x32
# )

model_space = ENASpace(
    width=16,           # the initial filters (channel number) for the model
    num_cells=8,        # the number of stacked cells in total
    dataset='cifar'     # to give a hint about input resolution, here is 32x32
)




config = RetiariiExeConfig(execution_engine='oneshot')
experiment = RetiariiExperiment(model_space, evaluator=evaluator, strategy=strategy)
experiment.run(config)
# trategy.run()
exported_arch = experiment.export_top_models('dict')[0]
print(exported_arch)
# strategy.export_top_models()



with fixed_arch(exported_arch):
    # final_model = DartsSpace(width=16, num_cells=8, dataset='cifar')
    final_model = ENASpace(width=16, num_cells=8, dataset='cifar')
train_loader = DataLoader(train_data, batch_size=96, num_workers=6)  # Use the original training data


max_epochs = 1

evaluator = Classification(
    learning_rate=1e-3,
    weight_decay=1e-4,
    train_dataloaders=train_loader,
    val_dataloaders=search_valid_loader,
    max_epochs=max_epochs,
    gpus=1,
    export_onnx=False,          # Disable ONNX export for this experiment
    fast_dev_run=fast_dev_run   # Should be false for fully training
)

evaluator.fit(final_model)



this is my enas code,I don't know if there is any error while it can run. I paste it here for these in need.

**Comment by Drxdx** [comment]
> ```python
> ```python
> import nni
> import numpy as np
> from nni.retiarii.evaluator.pytorch import Classification
> from torch.utils.data import SubsetRandomSampler
> from torchvision import transforms
> from torchvision.datasets import CIFAR10
> from nni.retiarii.evaluator.pytorch import DataLoader
> 
> 
> from nni.retiarii.experiment.pytorch import RetiariiExperiment, RetiariiExeConfig
> from nni.retiarii.strategy.rl  import PolicyBasedRL
> from nni.retiarii import fixed_arch
> 
> from nni.retiarii.strategy import ENAS as ENAStrategy
> from nni.retiarii.hub.pytorch import ENAS as ENASpace
> 
> from nni.retiarii.strategy import DARTS as DartsStrategy
> from nni.retiarii.hub.pytorch import DARTS as DartsSpace
> 
> 
> CIFAR_MEAN = [0.49139968, 0.48215827, 0.44653124]
> CIFAR_STD = [0.24703233, 0.24348505, 0.26158768]
> 
> # fast_dev_run = True
> fast_dev_run = False
> 
> # strategy = DartsStrategy()
> strategy = ENAStrategy(reward_metric_name='val_acc')
> 
> transform = transforms.Compose([
>     transforms.RandomCrop(32, padding=4),
>     transforms.RandomHorizontalFlip(),
>     transforms.ToTensor(),
>     transforms.Normalize(CIFAR_MEAN, CIFAR_STD),
> ])
> 
> train_data = nni.trace(CIFAR10)(root='./data', train=True, download=True, transform=transform)
> 
> num_samples = len(train_data)
> indices = np.random.permutation(num_samples)
> split = num_samples // 2
> 
> search_train_loader = DataLoader(
>     train_data, batch_size=64, num_workers=6,
>     sampler=SubsetRandomSampler(indices[:split]),
> )
> 
> search_valid_loader = DataLoader(
>     train_data, batch_size=64, num_workers=6,
>     sampler=SubsetRandomSampler(indices[split:]),
> )
> 
> evaluator = Classification(
>     learning_rate=1e-3,
>     weight_decay=1e-4,
>     train_dataloaders=search_train_loader,
>     val_dataloaders=search_valid_loader,
>     # max_epochs=10,
>     max_epochs=1,
>     gpus=1,
>     fast_dev_run=fast_dev_run,
> )
> 
> # model_space = DartsSpace(
> #     width=16,           # the initial filters (channel number) for the model
> #     num_cells=8,        # the number of stacked cells in total
> #     dataset='cifar'     # to give a hint about input resolution, here is 32x32
> # )
> 
> model_space = ENASpace(
>     width=16,           # the initial filters (channel number) for the model
>     num_cells=8,        # the number of stacked cells in total
>     dataset='cifar'     # to give a hint about input resolution, here is 32x32
> )
> 
> 
> 
> 
> config = RetiariiExeConfig(execution_engine='oneshot')
> experiment = RetiariiExperiment(model_space, evaluator=evaluator, strategy=strategy)
> experiment.run(config)
> # trategy.run()
> exported_arch = experiment.export_top_models('dict')[0]
> print(exported_arch)
> # strategy.export_top_models()
> 
> 
> 
> with fixed_arch(exported_arch):
>     # final_model = DartsSpace(width=16, num_cells=8, dataset='cifar')
>     final_model = ENASpace(width=16, num_cells=8, dataset='cifar')
> train_loader = DataLoader(train_data, batch_size=96, num_workers=6)  # Use the original training data
> 
> 
> max_epochs = 1
> 
> evaluator = Classification(
>     learning_rate=1e-3,
>     weight_decay=1e-4,
>     train_dataloaders=train_loader,
>     val_dataloaders=search_valid_loader,
>     max_epochs=max_epochs,
>     gpus=1,
>     export_onnx=False,          # Disable ONNX export for this experiment
>     fast_dev_run=fast_dev_run   # Should be false for fully training
> )
> 
> evaluator.fit(final_model)
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> this is my enas code,I don't know if there is any error while it can run. I paste it here for these in need.
> ```

good job!
